scapy/contrib/isotp.py

```diff
@@ -1611,3 +1613,370 @@ def recv(self, x=0xffff):
+    return return_values
+
+
+def ISOTPScan(sock, scan_range=range(0x7ff + 1), extended_addressing=False,
+              noise_listen_time=2,
+              sniff_time=0.1,
+              output_format=None,
+              can_interface="can0",
+              verbose=False):
+
+    """Scan for ISOTP Sockets on a bus and return findings
+
+    Args:
+            sock: CANSocket object to communicate with the bus under scan
+            scan_range: hexadecimal range of CAN-Identifiers to scan.
+                        Default is 0x0 - 0x7ff
+            extended_addressing: scan with ISOTP extended addressing
+            noise_listen_time: seconds to listen for default
+                               communication on the bus
+            sniff_time: time the scan waits for isotp flow control responses
+                        after sending a first frame
+            output_format: defines the format of the returned
+                           results (text, code or sockets). Provide a string
+                           e.g. "text". Default is "socket".
+            can_interface: interface used to create the returned code/sockets
+            verbose: displays information during scan
+
+    Scan for ISOTP Sockets in the defined range and returns found sockets
+    in a specified format. The format can be:
+    text: human readable output
+    code: python code for copy&paste
+    sockets: if output format is not specified, ISOTPSockets will be
+             created and returned in a list
+    """
+
+    if verbose:
+        print("Filtering background noise...")
+
+    # Send dummy packet. In most cases, this triggers activity on the bus.
+    dummy_pkt = CAN(identifier=0x123,
+                    data=b'\xaa\xbb\xcc\xdd\xee\xff\xaa\xbb')
+
+    background_pkts = sock.sniff(timeout=noise_listen_time,
+                                 started_callback=lambda:
+                                 sock.send(dummy_pkt))
+
+    noise_ids = list(set(pkt.identifier for pkt in background_pkts))
+
+    if extended_addressing:
+        found_packets = scan_extended(sock, scan_range, noise_ids=noise_ids,
+                                      sniff_time=sniff_time, verbose=verbose)
+    else:
+        found_packets = scan(sock, scan_range, noise_ids=noise_ids,
+                             sniff_time=sniff_time, verbose=verbose)
+
+    filter_periodic_packets(found_packets, verbose)
+
+    if output_format == "text":
+        return generate_text_output(found_packets)
+    if output_format == "code":
+        return generate_code_output(found_packets, can_interface)
+    return generate_isotp_list(found_packets, can_interface)
+
+
+def generate_text_output(found_packets):
+    if not found_packets:
+        return "No packets found."
+
+    text = "\nFound %s ISOTP-FlowControl Packet(s):" % len(found_packets)
+    for pack in found_packets:
+        extended_id = pack > 0x7ff
+        if extended_id:
+            send_id = int(pack / 256)
```

Comment:
You could use the `//` operator to do it in one operation

Reply:
done